source/cbor.c

```diff
@@ -0,0 +1,708 @@
+#include "external/libcbor/cbor.h"
+#include <aws/common/cbor.h>
+
+#include <aws/common/array_list.h>
+#include <aws/common/logging.h>
+#include <aws/common/private/external_module_impl.h>
+#include <aws/common/ref_count.h>
+#include <fenv.h>
+#include <float.h>
+#include <inttypes.h>
+#include <math.h>
+
+static bool s_aws_cbor_module_initialized = false;
+
+void aws_cbor_module_init(struct aws_allocator *allocator) {
+    (void)allocator;
+    if (!s_aws_cbor_module_initialized) {
+        /* Not allow any allocation from libcbor */
+        cbor_set_allocs(NULL, NULL, NULL);
+        s_aws_cbor_module_initialized = true;
+    }
+}
+
+void aws_cbor_module_cleanup(void) {
+    if (s_aws_cbor_module_initialized) {
+        s_aws_cbor_module_initialized = false;
+    }
+}
+
+/*******************************************************************************
+ * ENCODE
+ ******************************************************************************/
+
+struct aws_cbor_encoder {
+    struct aws_allocator *allocator;
+
+    /* Refcount maybe */
+    struct aws_ref_count refcount;
+    struct aws_byte_buf encoded_buf;
+};
+
+struct aws_cbor_encoder *aws_cbor_encoder_new(struct aws_allocator *allocator, size_t init_buffer_size) {
```

Comment:
would users know whats a good size they can pass in for init buffer size. maybe we just pick default for them

Reply:
another thing to consider is should we allow them to pass a buffer they have already allocated?